### ISSUE TITLE
Bug: unstable_useMutableSource throws error when mutated before subscribe

### ISSUE THREAD

**Issue by SreeniIO** [facebook/react#19948]
React version: 0.0.0-experimental-94c0244ba

## Steps To Reproduce

Mutating the external source during the initial render before subscribe throws error. 

https://codesandbox.io/s/usemutablesource-7nbs6?file=/src/App.js

![image](https://user-images.githubusercontent.com/8140606/94963407-f7033280-04ac-11eb-9cf7-e00c6630f55e.png)

1. Navigate to the above code sandbox
2. Click on the Mutate button. Final Value is rendered without any error. (Success Mode)
3. Toggle to Error Mode and click on the Mutate button again.
4. Error thrown (and the UI is rendered with the Final Value when run from vscode)

> Note: In our actual codebase (private), we are seeing this error at different instances e.g. component getting updated due to `setState` call in `useLayoutEffect` etc.

## The current behavior

Seeing errors in development build.

![image](https://user-images.githubusercontent.com/8140606/94962796-e3a39780-04ab-11eb-82d9-b24e4eaa5cc1.png)

## The expected behavior

These errors thrown by useMutableSource should not showup on the UI and not passed to the error boundary as the component is eventually re-rendered with the latest state without any error.

**Comment by SreeniIO** [comment]
I'm using this workaround for now
```javascript
  try {
    value = useMutableSource(mutableSource, getSnapshot, defaultSubscribe);
  } catch (error) {
    throw new Promise((resolve) => resolve());
  }
```
https://codesandbox.io/s/usemutablesource-forked-0suof?file=/src/App.js:1126-1269

**Comment by bvaughn** [comment]
This is interesting. The following reduced test case passes when I run it locally using the test renderer:
```js
fit('19948', () => {
  const source = createSource('initial');
  const mutableSource = createMutableSource(source, param => param.version);

  function Component() {
    const value = useMutableSource(mutableSource, defaultGetSnapshot, defaultSubscribe);
    Scheduler.unstable_yieldValue('Component:' + value);
    if (value === 'initial') {
      source.value = 'updated';
    }
    return <div>value: {value}</div>;
  }

  const ReactDOM = require('react-dom');
  const TestUtils = require('react-dom/test-utils');

  const container = document.createElement('div');
  act(() => {
    ReactDOM.render(<Component />, container);

    // Using the createRoot() API also passes.
    // ReactDOM.unstable_createRoot(container).render(<Component />);
  });

  expect(Scheduler).toHaveYielded(['Component:initial', 'Component:updated']);
  expect(container.textContent).toEqual("value: updated");
});
```

But throws the error you've reported when run [on Code Sandbox](https://codesandbox.io/s/practical-jang-y9qcs?file=/src/App.js).

**Comment by bvaughn** [comment]
Note that this test app also works when run with a UMD build:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <title>Hello World</title>
    <script src="https://unpkg.com/react@0.0.0-experimental-4ead6b530/umd/react.development.js"></script>
    <script src="https://unpkg.com/react-dom@0.0.0-experimental-4ead6b530/umd/react-dom.development.js"></script>

    <!-- Don't use this in production: -->
    <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
  </head>
  <body>
    <div id="root"></div>
    <script type="text/babel">
      const {
        unstable_createMutableSource: createMutableSource,
        unstable_useMutableSource: useMutableSource,
      } = React;

      function createSource(initialValue) {
        const callbacks = [];
        let revision = 0;
        let value = initialValue;
        return {
          subscribe(callback) {
            if (callbacks.indexOf(callback) < 0) {
              callbacks.push(callback);
            }
            return () => {
              const index = callbacks.indexOf(callback);
              if (index >= 0) {
                callbacks.splice(index, 1);
              }
            };
          },
          get listenerCount() {
            return callbacks.length;
          },
          set value(newValue) {
            revision++;
            value = newValue;
            callbacks.forEach(callback => callback());
          },
          get value() {
            return value;
          },
          get version() {
            return revision;
          },
        };
      }

      const getSnapshot = source => source.value;
      const subscribe = (source, callback) => source.subscribe(callback);

      const source = createSource('initial value');
      const mutableSource = createMutableSource(source, param => param.version);

      function Component() {
        const value = useMutableSource(mutableSource, getSnapshot, subscribe);
        console.log('<Component> value:', value);
        if (value === 'initial value') {
          source.value = 'updated value';
        }
        return <div>value: {value}</div>;
      }

      ReactDOM.render(
        <Component />,
        document.getElementById('root')
      );

    </script>
    <!--
      Note: this page is a great way to try React but it's not suitable for production.
      It slowly compiles JSX with Babel in the browser and uses a large development build of React.

      Read this section for a production-ready setup with JSX:
      https://reactjs.org/docs/add-react-to-a-website.html#add-jsx-to-a-project

      In a larger project, you can use an integrated toolchain that includes JSX instead:
      https://reactjs.org/docs/create-a-new-react-app.html

      You can also use React without JSX, in which case you can remove Babel:
      https://reactjs.org/docs/react-without-jsx.html
    -->
  </body>
</html>
```

The same error is thrown when running the example app with `create-react-app` though.

**Comment by bvaughn** [comment]
Looks like this check is causing the error to be thrown:
https://github.com/facebook/react/blob/6b28eb6175e5251b31507749b79c55c0e389b6ef/packages/react-reconciler/src/ReactFiberHooks.new.js#L891-L893

The first time the component is rendered, `currentRenderVersion === null` and `version === 0`. After the mutation, `currentRenderVersion === 0` and `version === 1`.

The difference between my unit tests and Code Sandbox / CRA app is this: The unit test only calls `readFromUnsubcribedMutableSource` once– the first time the component is rendered, when it initializes the internal snapshot value:
https://github.com/facebook/react/blob/6b28eb6175e5251b31507749b79c55c0e389b6ef/packages/react-reconciler/src/ReactFiberHooks.new.js#L969-L972

The Code Sandbox / CRA app app calls `readFromUnsubcribedMutableSource` twice. The same line of code calls it though. Not yet sure why.

**Comment by bvaughn** [comment]
I believe the cause of this DEV only bug has been narrowed down to `StrictMode` (which is on by default for `unstable_createRoot`) and the `debugRenderPhaseSideEffectsForStrictMode` feature flag.

**Comment by bvaughn** [comment]
@SreeniIO After looking at this repro for a couple of hours last week, I'm pretty sure it's specific to a mutation that happens *in the render method* (which is not supported). Your bug report suggests that in your actual app, a similar behavior is triggered by mutations occurring in other places:

> Note: In our actual codebase (private), we are seeing this error at different instances e.g. component getting updated due to setState call in useLayoutEffect etc.

Can you share another repro that doesn't involve mutation while a component is rendering?

**Comment by bvaughn** [comment]
Oh neat. I was able to catch this in a failing test after all.
```js
const source = createSource("initial");
const mutableSource = createMutableSource(source, (param) => param.version);

function Component() {
  const value = useMutableSource(
    mutableSource,
    defaultGetSnapshot,
    defaultSubscribe
  );
  if (value === "initial") {
    source.value = "updated";
  }
  return null; // <div>value: {value}</div>;
}

const container = document.createElement("div");
act(() => {
  ReactDOM.render(
    <React.StrictMode>
      <Component />
    </React.StrictMode>,
    container
  );
});
```

**Comment by bvaughn** [comment]
@SreeniIO Thanks for the repro case and discussion. I've landed a solution in #20665 and I'd suggest reconsidering the workaround you mentioned. My reasoning is this (copied from the PR description):

> Mutating a mutable source (or any external variable) during render **is not expected to be supported**. However this error message misattributes the mutation as a React bug.
>
> ### Should this use case be supported?
> This seems very similar [lazily initializing a ref](https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily) so it might be tempting to think it's okay. Mutable sources are external though, so modifying them during render is a side effect [which makes it unsafe and unsupported](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects). So while we could "support" this case in a very limited way, we shouldn't.
